staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+		}
+	}
+	// TODO: check metrics is non-nil?
+	if a.PodSpecExtractor == nil {
+		return fmt.Errorf("PodSpecExtractor required")
+	}
+	if a.Registry == nil {
+		return fmt.Errorf("Registry required")
+	}
+	if a.NamespaceGetter == nil {
+		return fmt.Errorf("NamespaceGetter required")
+	}
+	if a.PodLister == nil {
+		return fmt.Errorf("PodLister required")
+	}
+	return nil
+}
+
+// Validate admits an API request.
+// The objects in admission attributes are expected to be external v1 objects that we care about.
+func (a *Admission) Validate(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	var response admissionv1.AdmissionResponse
+	switch attrs.GetResource().GroupResource() {
+	case corev1.Resource("namespaces"):
+		response = a.ValidateNamespace(ctx, attrs)
+	case corev1.Resource("pods"):
+		response = a.ValidatePod(ctx, attrs)
+	default:
+		response = a.ValidatePodController(ctx, attrs)
+	}
+
+	// TODO: record metrics.
+
+	return response
+}
+
+func (a *Admission) ValidateNamespace(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	namespace := attrs.GetObject().(*corev1.Namespace)
+	var oldNamespace *corev1.Namespace
+	if attrs.GetOperation() == admission.Update {
+		var ok bool
+		oldNamespace, ok = attrs.GetOldObject().(*corev1.Namespace)
+		if !ok {
+			klog.InfoS("failed to assert old namespace type", "type", reflect.TypeOf(attrs.GetOldObject()))
+			return internalErrorResponse("failed to decode old namespace")
+		}
+	}
+
+	policy, err := a.PolicyToEvaluate(namespace.Labels)
+
+	// if update, check if policy labels changed
+	if oldNamespace != nil {
+		oldPolicy, oldErr := a.PolicyToEvaluate(oldNamespace.Labels)
+
+		// Only validate on diff.
+		if err != nil && err.Error() != oldErr.Error() {
+			return invalidResponse(err.Error())
+		}
+		// Skip validation if the policy is unchanged.
+		if policy.Enforce.Version == oldPolicy.Enforce.Version {
+			return allowedResponse()
+		}
+		// We can skip validation if the enforce policy is the same pinned version and relaxed.
+		if !policy.Enforce.Latest() && policy.Enforce.Version == oldPolicy.Enforce.Version &&
+			api.CompareLevels(policy.Enforce.Level, oldPolicy.Enforce.Level) < 1 {
+			return allowedResponse()
+		}
+	} else if err != nil { // Create
+		return invalidResponse(err.Error())
+	}
+
+	// We only need to test pods if the enforce policy is not privileged.
+	if policy.Enforce.Level == api.LevelPrivileged {
+		return allowedResponse()
+	}
+
+	response := allowedResponse()
+	response.Warnings = a.CheckPodsInNamespace(ctx, namespace.Name, policy.Enforce)
+
+	return response
+}
+
+func (a *Admission) ValidatePod(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	// short-circuit on subresources other than ephemeralcontainers
+	if attrs.GetSubresource() != "" && attrs.GetSubresource() != "ephemeralcontainers" {
+		return allowedResponse()
```

Comment:
This allows by default instead of fully enumerating and denying unknown?  If this is used as a library by a webhook which may skew levels, I would expect enumeration and deny-by-default.

Reply:
yeah, enumerating seems fine